### PR TITLE
Add signature v4

### PR DESCRIPTION
**Pull request by kevin-yen**
*Changes*
* Added signature v4 pre-signed URL function
* Extracted out current v2 signature process to a separate function
* Added `signature_version` keyword parameter to `s3_sign_url` to select which signature process version to use
* Set default signature version to v4
* Updated documentation

### CODE REVIEW

**Comment by kevin-yen** [comment]
https://github.com/JuliaCloud/AWSS3.jl/issues/89

File: src/AWSS3.jl [hunk 1]
```diff
@@ -800,6 +779,111 @@ function s3_sign_url(aws::AWSConfig, bucket, path, seconds=3600;
...
+url = s3_sign_url("my_bucket", "my_file.txt";
+                  verb="PUT", content_type="text/plain")
+
+Requests.put(URI(url), "Hello!";
+             headers=Dict("Content-Type" => "text/plain"))
+```
+"""
+function s3_sign_url(aws::AWSConfig, bucket, path, seconds=3600;
+                     verb="GET", content_type="application/octet-stream",
+                     protocol="http", signature_version="v4")
```
Comment (mattBrzezinski) [inline src/AWSS3.jl:872]: This is probably getting way too into it, but we could do what is below. However, it's probably overkill. 🤷 

```
export SIGNATURE_VERSION
@enum SIGNATURE_VERSION v2=0 v4=1

function s3_sign_url(
    ::Type{Val{v2}}, aws::AWSConfig, bucket, path, seconds=3600; 
    verb="GET", content_type="application/octet-stream", protocol="http",
)
    # v2 signing code here
end

function s3_sign_url(
    ::Type{Val{v4}}, aws::AWSConfig, bucket, path, seconds=3600; 
    verb="GET", content_type="application/octet-stream", protocol="http",
)
    # v4 signing code here
end
```

File: src/AWSS3.jl [hunk 2]
```diff
@@ -800,6 +779,111 @@ function s3_sign_url(aws::AWSConfig, bucket, path, seconds=3600;
...
+
+    if signature_version == "v2"
+        _s3_sign_url_v2(aws, bucket, path, seconds;
+                        verb = verb, content_type = content_type,
+                        protocol = protocol)
```
Comment (mattBrzezinski) [inline src/AWSS3.jl:877]:
```suggestion
        _s3_sign_url_v2(
            aws, bucket, path, seconds;
            verb=verb, content_type=content_type, protocol=protocol
        )
```

File: src/AWSS3.jl [hunk 3]
```diff
@@ -800,6 +779,111 @@ function s3_sign_url(aws::AWSConfig, bucket, path, seconds=3600;
...
+    elseif signature_version == "v4"
+        _s3_sign_url_v4(aws, bucket, path, seconds;
+                        verb = verb, content_type = content_type,
+                        protocol = protocol)
```
Comment (mattBrzezinski) [inline src/AWSS3.jl:881]:
```suggestion
        _s3_sign_url_v4(
            aws, bucket, path, seconds;
            verb=verb, content_type=content_type, protocol=protocol
        )
```

File: src/AWSS3.jl [hunk 4]
```diff
@@ -745,34 +745,13 @@ function s3_multipart_upload(aws::AWSConfig, bucket, path, io::IOStream,
...
-Requests.put(URI(url), "Hello!";
-             headers=Dict("Content-Type" => "text/plain"))
-```
-"""
-function s3_sign_url(aws::AWSConfig, bucket, path, seconds=3600;
-                     verb="GET", content_type="application/octet-stream",
-                     protocol="http")
+function _s3_sign_url_v2(aws::AWSConfig, bucket, path, seconds=3600;
+                         verb="GET", content_type="application/octet-stream",
+                         protocol="http")
```
Comment (mattBrzezinski) [inline src/AWSS3.jl:754]: Although this package doesn't officially support [BlueStyle](https://github.com/invenia/BlueStyle) I would like to move towards it.

```suggestion
function _s3_sign_url_v2(
    aws::AWSConfig, bucket, path, seconds=3600;
    verb="GET", content_type="application/octet-stream", protocol="http",
)
```

File: src/AWSS3.jl [hunk 5]
```diff
@@ -800,6 +779,111 @@ function s3_sign_url(aws::AWSConfig, bucket, path, seconds=3600;
...
+
+    service = "s3"
+    scheme = "AWS4"
+    algorithm = "HMAC-SHA256"
+    terminator = "aws4_request"
+
+    scope = date_stamp * "/" * aws[:region] * "/" * service * "/" * terminator
+    host = string("s3-", aws[:region], ".amazonaws.com")
+
+    headers = OrderedDict{String, String}("Host" => host)
```
[the remark this answers is not in the document]
Comment (kevin-yen) [reply src/AWSS3.jl:801]: In all examples I see for the header names, they all show them to be sorted
https://gist.github.com/kevin-yen/dd6968ab1640ab5212c6cf323941a5f4#file-aws4signerbase-java-L81 (from https://docs.aws.amazon.com/AmazonS3/latest/API/sig-v4-examples-using-sdks.html) 
image

**Comment by kevin-yen** [comment]
> Could you also write a couple of tests for this? Just checking that both v2 and v4 are working properly. If not I can add them in later this week.
>Just saw this after I had submitted a review. The Base.IOError is because that URL is only accessible by EC2 instances. I can take a look at what is happening tomorrow and post an update!
Yeah I can write couple of tests as soon as I can get the tests to pass locally

File: src/AWSS3.jl (hunk 4 above)
Comment (kevin-yen) [reply src/AWSS3.jl:754]: No problems! I was just trying to match the style of the rest of the file. I can go ahead and use BlueStyle

File: src/AWSS3.jl (hunk 1 above)
Comment (kevin-yen) [reply src/AWSS3.jl:872]: I have no opinion either way. I'd imagine we want to have a consistent style/format across all packages under JuliaCloud?

File: src/AWSS3.jl [hunk 6]
```diff
@@ -800,6 +779,111 @@ function s3_sign_url(aws::AWSConfig, bucket, path, seconds=3600;
...
+    now_datetime = now(Dates.UTC)
+    datetime_stamp = Dates.format(now_datetime, "YYYYmmddTHHMMSSZ")
+    date_stamp = Dates.format(now_datetime, "YYYYmmdd")
+
+    service = "s3"
+    scheme = "AWS4"
+    algorithm = "HMAC-SHA256"
+    terminator = "aws4_request"
+
+    scope = date_stamp * "/" * aws[:region] * "/" * service * "/" * terminator
```
[the remark this answers is not in the document]
Comment (kevin-yen) [reply src/AWSS3.jl:798]: Sure! Let me make that changes

File: src/AWSS3.jl [hunk 7]
```diff
@@ -745,34 +745,13 @@ function s3_multipart_upload(aws::AWSConfig, bucket, path, io::IOStream,
 end
 
 
+using DataStructures
```
[the remark this answers is not in the document]
Comment (kevin-yen) [reply src/AWSS3.jl:748]: I can go ahead and try. Do you remember how the issue will manifest? When I run a test? During precompile?

[the remark this answers is not in the document]
Comment (mattBrzezinski) [reply src/AWSS3.jl]: I believe you would get a compilation error, however I'm not quite sure anymore.

I guess this is a perfect example of why comments are nice to have :P

File: src/AWSS3.jl (hunk 5 above)
[the remark this answers is not in the document]
Comment (mattBrzezinski) [reply src/AWSS3.jl:801]: Hmm interesting, I feel like the order shouldn't matter for the canonical query string, canonical headers, and signed headers. Do you mind trying this with a `Dict()` and seeing if that works? I would myself but this PR is going through a fork so I wouldn't be able to push changes easily anyways.

File: src/AWSS3.jl [hunk 8]
```diff
@@ -800,6 +778,122 @@ function s3_sign_url(aws::AWSConfig, bucket, path, seconds=3600;
...
+function s3_sign_url(
+    aws::AWSConfig, bucket, path, seconds=3600;
+    verb="GET", content_type="application/octet-stream", protocol="http",
+    signature_version="v4"
+)
+
+    if signature_version == "v2"
+        _s3_sign_url_v2(
+            aws, bucket, path, seconds;
+            verb = verb, content_type = content_type, protocol = protocol
```
Comment (mattBrzezinski) [inline src/AWSS3.jl:885]:
```suggestion
            verb=verb, content_type=content_type, protocol=protocol
```

File: src/AWSS3.jl [hunk 9]
```diff
@@ -800,6 +778,122 @@ function s3_sign_url(aws::AWSConfig, bucket, path, seconds=3600;
...
+        )
+    elseif signature_version == "v4"
+        _s3_sign_url_v4(
+            aws, bucket, path, seconds;
+            verb = verb, content_type = content_type, protocol = protocol
```
Comment (mattBrzezinski) [inline src/AWSS3.jl:890]:
```suggestion
            verb=verb, content_type=content_type, protocol=protocol
```

File: src/AWSS3.jl [hunk 10]
```diff
@@ -800,6 +778,122 @@ function s3_sign_url(aws::AWSConfig, bucket, path, seconds=3600;
...
+    datetime_stamp = Dates.format(now_datetime, "YYYYmmddTHHMMSSZ")
+    date_stamp = Dates.format(now_datetime, "YYYYmmdd")
+
+    service = "s3"
+    scheme = "AWS4"
+    algorithm = "HMAC-SHA256"
+    terminator = "aws4_request"
+
+    scope = "$date_stamp/$(aws[:region])/$service/$terminator"
+    host = string("s3-", aws[:region], ".amazonaws.com")
```
Comment (mattBrzezinski) [inline src/AWSS3.jl:799]:
```suggestion
    host = "s3-$(aws[:region]).amazonaws.com"
```

File: src/AWSS3.jl [hunk 11]
```diff
@@ -800,6 +778,122 @@ function s3_sign_url(aws::AWSConfig, bucket, path, seconds=3600;
     return "$endpoint/$path?$(HTTP.escapeuri(query))"
 end
 
+
+function _s3_sign_url_v4(
+    aws::AWSConfig, bucket, path, seconds=3600;
+    verb="GET", content_type="application/octet-stream", protocol="http"
```
Comment (mattBrzezinski) [inline src/AWSS3.jl:784]:
```suggestion
    verb="GET", content_type="application/octet-stream", protocol="http",
```

File: src/AWSS3.jl [hunk 12]
```diff
@@ -800,6 +778,122 @@ function s3_sign_url(aws::AWSConfig, bucket, path, seconds=3600;
...
+                  verb="PUT", content_type="text/plain")
+
+Requests.put(URI(url), "Hello!";
+             headers=Dict("Content-Type" => "text/plain"))
+```
+"""
+function s3_sign_url(
+    aws::AWSConfig, bucket, path, seconds=3600;
+    verb="GET", content_type="application/octet-stream", protocol="http",
+    signature_version="v4"
```
Comment (mattBrzezinski) [inline src/AWSS3.jl:879]:
```suggestion
    signature_version="v4",
```

File: src/AWSS3.jl [hunk 13]
```diff
@@ -800,6 +778,122 @@ function s3_sign_url(aws::AWSConfig, bucket, path, seconds=3600;
...
+    signature = digest(MD_SHA256, string_to_sign, key_signing)
+
+    query["X-Amz-Signature"] = signature |> bytes2hex
+
+    return string(protocol, "://", host, path, "?", HTTP.escapeuri(query))
+end
+
+
+"""
+    s3_sign_url([::AWSConfig], bucket, path, [seconds=3600];
```
Comment (mattBrzezinski) [inline src/AWSS3.jl:852]: Missing `protocol`

File: src/AWSS3.jl [hunk 14]
```diff
@@ -744,35 +745,12 @@ function s3_multipart_upload(aws::AWSConfig, bucket, path, io::IOStream,
...
-Requests.put(URI(url), "Hello!";
-             headers=Dict("Content-Type" => "text/plain"))
-```
-"""
-function s3_sign_url(aws::AWSConfig, bucket, path, seconds=3600;
-                     verb="GET", content_type="application/octet-stream",
-                     protocol="http")
+function _s3_sign_url_v2(
+    aws::AWSConfig, bucket, path, seconds=3600;
+    verb="GET", content_type="application/octet-stream", protocol="http"
```
Comment (mattBrzezinski) [inline src/AWSS3.jl:752]:
```suggestion
    verb="GET", content_type="application/octet-stream", protocol="http",
```

File: src/AWSS3.jl (hunk 10 above)
Comment (mattBrzezinski) [inline src/AWSS3.jl:799]:
```suggestion
    host = string("s3-", aws[:region], ".amazonaws.com")
```

I'm not sure if `s3-` is correct here, should this not be `s3.`?

Comment (kevin-yen) [reply src/AWSS3.jl:799]: This is from the Java sample code I relied upon
https://gist.github.com/kevin-yen/32183655ad7e0b57bec7d678a51e8260#file-presignedurlsample-java-L32
so I think it's `s3-[region]` for regions outside of us-east-1. I had to double-check that but I believe it is `s3-` and I have it tested the URL in our qa and prod environments so I'm fairly positive it's correct

Comment (mattBrzezinski) [reply src/AWSS3.jl:799]: I pulled down this branch and ran the tests locally and is was problematic on this [line](https://github.com/JuliaCloud/AWSS3.jl/blob/master/test/awss3.jl#L73). I think adding in tests to the package will clear this up.

File: src/AWSS3.jl (hunk 7 above)
[the remark this answers is not in the document]
Comment (kevin-yen) [reply src/AWSS3.jl]: Haha true. I can try it and see if I run into problems

File: src/AWSS3.jl (hunk 5 above)
[the remark this answers is not in the document]
Comment (kevin-yen) [reply src/AWSS3.jl:801]: Yeah I can try it

File: src/AWSS3.jl (hunk 10 above)
Comment (kevin-yen) [reply src/AWSS3.jl]: I'm getting a different error but I just realized I haven't set up the AWS credentials on the machine I was running the test. I think that might be the issue. Let me get the right credentials on to the machine

Comment (kevin-yen) [reply src/AWSS3.jl]: Yep that was it. I needed to setup AWS credentials on my local computer

**Comment by mattBrzezinski** [comment]
bors try

**Comment by mattBrzezinski** [comment]
bors try

**Comment by mattBrzezinski** [comment]
This repo is having the same issues as [AWSSQS.jl](https://github.com/JuliaCloud/AWSSQS.jl) did, I can resolve this another time. But this looks good to go!

**Pull request merged**